photon-client/src/main/scala/com/linkedin/photon/ml/data/avro/AvroDataWriter.scala

```diff
@@ -0,0 +1,151 @@
+
+      var responseType: Option[DataType] = None
+      var offsetType: Option[DataType] = None
+      var weightType: Option[DataType] = None
+
+      rows.map { r: Row =>
+        responseType match {
+          case None =>
+            responseType = Some(r.schema(responseColumn).dataType)
+            if (hasOffset) {
+              offsetType = Some(r.schema("offset").dataType)
+            }
+            if (hasWeight) {
+              weightType = Some(r.schema("weight").dataType)
+            }
+          case Some(_) =>
+        }
+        val features = r.getAs[Vector](featureColumn)
+        val response = getValueAsDouble(r, responseColumn, responseType.get)
+        val offset = offsetType match {
+          case Some(t) => getValueAsDouble(r, "offset", t)
+          case None => 0.0D
+        }
+        val weight = weightType match {
+          case Some(t) => getValueAsDouble(r, "weight", t)
+          case None => 1.0D
+        }
+        rowBuilder
+          .setResponse(response)
+          .setOffset(offset)
+          .setWeight(weight)
+          .setFeatures(buildAvroFeatures(features, broadcastReverseMap.value))
+          .setRandomTestField("AvroWriter Generated")
+          .build()
+      }
+    }
+
+    // Write the converted dataset back to HDFS
+    // If the output paths already exist, remove them
+    val fs = FileSystem.get(sc.hadoopConfiguration)
+    val output = new Path(outputPath)
+    if (fs.exists(output)) {
+      fs.delete(output, true)
+    }
+
+    AvroUtils.saveAsAvro[SimplifiedResponsePrediction](
+      avroDataset,
+      outputPath,
+      SimplifiedResponsePrediction.getClassSchema.toString)
+  }
+
+  /**
+   * Helper function to convert Row index field to double
+   *
+   * @param row A training record in [[Row]] format
+   * @param fieldName The index of particular field
+   * @param dataType The dataype of this field
+   * @return A double in this field
+   */
+  protected[data] def getValueAsDouble(row: Row, fieldName: String, dataType: DataType): Double = {
+
+    dataType match {
+      case IntegerType => row.getAs[Int](fieldName).toDouble
+      case LongType => row.getAs[Long](fieldName).toDouble
+      case FloatType => row.getAs[Float](fieldName).toDouble
+      case BooleanType => if (row.getAs[Boolean](fieldName)) 1.0D else 0.0D
+      case DoubleType => row.getAs[Double](fieldName)
+      case _ =>
+        throw new IllegalArgumentException(s"Unsupported data type $dataType")
+    }
+  }
+
+  /**
+   * Build a list of Avro Feature instances for the given list [[Vector]] and [[IndexMap]]
+   *
+   * @param vector The extracted feature in [[Vector]] for a particular training instance
+   * @param indexMap The reverse index map from feature to index
+   * @return A list of Avro Feature instances built from the vector
+   */
+  protected[data] def buildAvroFeatures(vector: Vector, indexMap: Map[Int, String]): java.util.List[FeatureAvro] = {
```

Comment:
You can use Scala data strucures instead here, and then convert to Java at the end when needed, e.g.:

```scala
import scala.collection.JavaConverters._
List(1,2,3).asJava
```

Reply:
done